src/replication.c

```diff
@@ -567,47 +651,50 @@ handle_read_resp(spec_t *spec, replica_t *replica)
 	rcmd_t *rep_cmd = NULL;
 	rcommon_cmd_t *rcommq_ptr = NULL;
 	zvol_io_hdr_t *io_rsp_hdr = replica->io_resp_hdr;
-	void *data = replica->io_resp_data;
 	uint64_t datalen = 0;
 	uint8_t *dataptr = NULL;
-
+	data_read_t *data_read = (data_read_t *)malloc(sizeof(data_read_t));
 	MTX_LOCK(&replica->r_mtx);
 	//Find IO in read queue, signal luworker, and dequeue
 	TAILQ_FOREACH(rep_cmd, &replica->read_waitq, rread_cmd_next) {
 		if(io_rsp_hdr->io_seq == rep_cmd->io_seq) {
+			rcommq_ptr = rep_cmd->rcommq_ptr;
 			io_found = true;
 			rep_cmd->ack_recvd = true;
 			TAILQ_REMOVE(&replica->read_waitq, rep_cmd, rread_cmd_next);
 			MTX_UNLOCK(&replica->r_mtx);
-			if (io_rsp_hdr->status == ZVOL_OP_STATUS_OK)
-				dataptr = get_read_resp_data(io_rsp_hdr, data, &datalen);
+
+			MTX_LOCK(&spec->rq_mtx);
+			MTX_UNLOCK(&spec->rq_mtx);
+
 			MTX_LOCK(&spec->rcommonq_mtx);
-			rcommq_ptr = rep_cmd->rcommq_ptr;
 			MTX_LOCK(&rcommq_ptr->rcommand_mtx);
-			if (io_rsp_hdr->status == ZVOL_OP_STATUS_OK)
-				rcommq_ptr->status = 1;
-			else
-				rcommq_ptr->status = -1;
-			rcommq_ptr->data = dataptr;
-			rcommq_ptr->data_len = datalen;
-			luworker_id = rcommq_ptr->luworker_id;
-			put_to_mempool(&rcmd_mempool, rep_cmd);
-
-			signal_luworker();
+			if (rep_cmd->status == ZVOL_OP_STATUS_OK) {
+				rcommq_ptr->acks_recvd++;
+			} else {
+				rcommq_ptr->ios_aborted++;
+			}
+			data_read->io_resp_data = data_read->io_resp_data_ptr_cpy = replica->io_resp_data;
+			data_read->bytes_consumed = 0;
 
-			TAILQ_REMOVE(&spec->rcommon_waitq, rcommq_ptr, wait_cmd_next);
-			rcommq_ptr->state = CMD_EXECUTION_DONE;
-			if(rcommq_ptr->completed) {
-				MTX_UNLOCK(&rcommq_ptr->rcommand_mtx);
+			TAILQ_INSERT_TAIL(&rcommq_ptr->data_read_ptr, data_read, data_next);
+			luworker_id = rcommq_ptr->luworker_id;
+			if(!rcommq_ptr->send_multiple_read) {
+				handle_read_healthy();
+			} else if (rcommq_ptr->acks_recvd == MAX(spec->replica_count - spec->consistency_factor + 1,
+								spec->consistency_factor)) {
+				handle_read_consistency_met();
 				MTX_UNLOCK(&spec->rcommonq_mtx);
-				clear_rcomm_cmd(rcommq_ptr);
+				MTX_UNLOCK(&rcommq_ptr->rcommand_mtx);
+			} else if (rcommq_ptr->acks_recvd + rcommq_ptr->ios_aborted
```

Comment:
we can just have the case for 'acks_recvd + ios_aborted == copies_sent'. In that also, we should be able to handle by going through all the responses received.
Reason for this approach is: There might be change in the state of replica after sending the requests to replicas, like, after sending the request, replica's rebuilding might have got completed and may not be sending metadata. So, after receiving all the responses, do:
- check for any response with io_num as 0, and send that as the response
- else, check whether acks_recvd is equal to MAX(.....), and build the response and send that
- else, send error response
Finally, do the cleanup of the command and its memory.

Reply:
If a response has io_num as 0 in the IO header, how do we find that IO in our queues. I went through the zrepl code, it looks like we need to add one more variable to our IO header to use this logic or use the flag variable.